From: Stained Glass OS <ke7oxh@gmail.com>
Subject: [PATCH] msvcp60: Implement _Lockit and locale::id assignment.

msvcp60's _Lockit::operator= (the VC 6 _Lockit has no data: nothing to
copy) and locale::id::operator= (copies the id).
---
 dlls/msvcp60/msvcp60.spec | 12 ++++++------
 dlls/msvcp90/locale.c     | 14 ++++++++++++++
 dlls/msvcp90/misc.c       | 12 ++++++++++++
 3 files changed, 32 insertions(+), 6 deletions(-)

diff --git a/dlls/msvcp60/msvcp60.spec b/dlls/msvcp60/msvcp60.spec
index 29553eb..b1a4c23 100644
--- a/dlls/msvcp60/msvcp60.spec
+++ b/dlls/msvcp60/msvcp60.spec
@@ -1577,9 +1577,9 @@
 @ stub -arch=arm ??4_Locinfo@std@@QAAAAV01@ABV01@@Z
 @ stub -arch=i386 ??4_Locinfo@std@@QAEAAV01@ABV01@@Z
 @ stub -arch=win64 ??4_Locinfo@std@@QEAAAEAV01@AEBV01@@Z
-@ stub -arch=arm ??4_Lockit@std@@QAAAAV01@ABV01@@Z
-@ stub -arch=i386 ??4_Lockit@std@@QAEAAV01@ABV01@@Z
-@ stub -arch=win64 ??4_Lockit@std@@QEAAAEAV01@AEBV01@@Z
+@ cdecl -arch=arm ??4_Lockit@std@@QAAAAV01@ABV01@@Z(ptr ptr) _Lockit_op_assign
+@ thiscall -arch=i386 ??4_Lockit@std@@QAEAAV01@ABV01@@Z(ptr ptr) _Lockit_op_assign
+@ cdecl -arch=win64 ??4_Lockit@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) _Lockit_op_assign
 @ cdecl -arch=arm ??4_Num_base@std@@QAAAAU01@ABU01@@Z(ptr ptr) std_Num_base_op_assign
 @ thiscall -arch=i386 ??4_Num_base@std@@QAEAAU01@ABU01@@Z(ptr ptr) std_Num_base_op_assign
 @ cdecl -arch=win64 ??4_Num_base@std@@QEAAAEAU01@AEBU01@@Z(ptr ptr) std_Num_base_op_assign
@@ -1613,9 +1613,9 @@
 @ cdecl -arch=arm ??4domain_error@std@@QAAAAV01@ABV01@@Z(ptr ptr) MSVCP_domain_error_assign
 @ thiscall -arch=i386 ??4domain_error@std@@QAEAAV01@ABV01@@Z(ptr ptr) MSVCP_domain_error_assign
 @ cdecl -arch=win64 ??4domain_error@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) MSVCP_domain_error_assign
-@ stub -arch=arm ??4id@locale@std@@QAAAAV012@ABV012@@Z
-@ stub -arch=i386 ??4id@locale@std@@QAEAAV012@ABV012@@Z
-@ stub -arch=win64 ??4id@locale@std@@QEAAAEAV012@AEBV012@@Z
+@ cdecl -arch=arm ??4id@locale@std@@QAAAAV012@ABV012@@Z(ptr ptr) locale_id_op_assign
+@ thiscall -arch=i386 ??4id@locale@std@@QAEAAV012@ABV012@@Z(ptr ptr) locale_id_op_assign
+@ cdecl -arch=win64 ??4id@locale@std@@QEAAAEAV012@AEBV012@@Z(ptr ptr) locale_id_op_assign
 @ cdecl -arch=arm ??4ios_base@std@@QAAAAV01@ABV01@@Z(ptr ptr) ios_base_assign
 @ thiscall -arch=i386 ??4ios_base@std@@QAEAAV01@ABV01@@Z(ptr ptr) ios_base_assign
 @ cdecl -arch=win64 ??4ios_base@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) ios_base_assign
diff --git a/dlls/msvcp90/locale.c b/dlls/msvcp90/locale.c
index d618e5c..3920ff2 100644
--- a/dlls/msvcp90/locale.c
+++ b/dlls/msvcp90/locale.c
@@ -16904,3 +16904,17 @@ unsigned int __cdecl time_get_short__Getcat_old(const locale_facet **facet)
     return time_get_short__Getcat(facet, locale_classic());
 }
 #endif
+
+#if _MSVCP_VER < 70
+/* ??4id@locale@std@@QAEAAV012@ABV012@@Z */
+/* ??4id@locale@std@@QEAAAEAV012@AEBV012@@Z */
+DEFINE_THISCALL_WRAPPER(locale_id_op_assign, 8)
+locale_id* __thiscall locale_id_op_assign(locale_id *this, const locale_id *rhs)
+{
+    TRACE("(%p %p)\n", this, rhs);
+#ifndef SG_MUTANT_LOCALE_ID_ASSIGN_NOP
+    this->id = rhs->id;
+#endif
+    return this;
+}
+#endif
diff --git a/dlls/msvcp90/misc.c b/dlls/msvcp90/misc.c
index 2386edd..2075eae 100644
--- a/dlls/msvcp90/misc.c
+++ b/dlls/msvcp90/misc.c
@@ -2285,3 +2285,15 @@ DWORD __cdecl _GetCurrentThreadId(void)
     return GetCurrentThreadId();
 }
 #endif
+
+#if _MSVCP_VER < 70
+/* ??4_Lockit@std@@QAEAAV01@ABV01@@Z */
+/* ??4_Lockit@std@@QEAAAEAV01@AEBV01@@Z */
+/* the VC 6 _Lockit has no data: assignment does nothing */
+DEFINE_THISCALL_WRAPPER(_Lockit_op_assign, 8)
+_Lockit* __thiscall _Lockit_op_assign(_Lockit *this, const _Lockit *rhs)
+{
+    TRACE("(%p %p)\n", this, rhs);
+    return this;
+}
+#endif
